README.md
# <img src="src/assets/logo.png" data-canonical-src="https://gyazo.com/eb5c5741b6a9a16c692170a41a49c858.png" height="30" />Flask

Sound for Improv and Theatre

![Flask screenshot](https://www.brandoncathcart.com/flask/img/Screenshot.JPG "Flask screenshot")

## Latest Release
Latest Stable Release: [v1.0.0-rc](https://github.com/brandoshizzle/REACTion/releases/latest)

[Download for Windows and OSX](https://www.brandoncathcart.com/flask/#download)


## Basic Instructions

* Drag sounds onto a key or the playlist to add it to the program
* Press the key to play the sound, press it again to stop
* Press space to play the first item of the playlist (highlighted in yellow)
* Right-click on a sound to bring up its settings
* Click on a sound to load its waveform
* Drag the edge of the waveform to adjust start/end points
* Drag playlist items to reorder, CTRL+C/X and CTRL+V to move sounds around on the keyboards
* Press DELETE when a song is selected to delete it
* F1-F8 to switch pages

## Changelog
See full Changelog [here](CHANGELOG.md)

## 3rd party libraries

Flask is lovingly built with these amazing projects:
* [Electron](http://electron.atom.io/)
* [Materialize](http://materializecss.com/)
* [Howler](https://github.com/goldfire/howler.js)
* [wavesurfer.js](https://github.com/katspaugh/wavesurfer.js)
* [Sortable](https://github.com/RubaXa/Sortable)
* [Jeditable](https://github.com/tuupola/jquery_jeditable)
* [jQuery](https://jquery.com/)

## License

Copyright 2018 Brandon Cathcart

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2019-12-21

### Added

-   Tutorial to take you through the basic features
-   Save and load shows
-   Select and edit multiple keys at once
-   Waveform start and end time handles! - no Ctrl or Cmd required
-   Separate playlist and pages fade times
-   Autoplay switch for playlist
-   Shuffle playlist
-   Splash screen
-   Edit icon next to sound/page names
-   Import REACTion and Flask v0.X.X projects
-   Flask Pro active for imported legacy projects
-   14-day free trial of Flask Pro

### Changed

-   Pro and free versions - free version limits pages and playlist
-   Panic switch now has confirmation (double tap)
-   Color picker in sound settings
-   Fade in time now dropdown - much more intuitive
-   Main settings redesigned
-   Ctrl/Cmd + P opens the settings
-   Ctrl/Cmd + S tells you that Flask is always saving
-   Visual updates
-   Updated to Electron v4.1.4
-   Updated Sortable JS to v1.8.4
-   Updated Materialize to v1.0.0

### Fixed

-   Next playlist waveform doesn't load if another sound is playing
-   Cutting or deleting a sound stops it from playing
-   Links all open in default OS browser

## [0.5.0] - 2019-01-13

### Added

-   Panic switch (Global stop) by pressing ESCAPE
-   Solo sound options - choose to auto-stop playing sounds when a new sound starts
-   Volume control sound in sound settings
-   Double-click volume bar to reset to 100%
-   Show changelog on first time after update

### Fixed

-   Fixed issue with playlist not switching song after fade out
-   Fixed issue with loaded bar not hiding when errors loading songs
-   Removed missed restricted characters from song ids
-   General refactoring and cleaning of code

### Changed

-   Name (Wowee!)
-   Misc fonts/logos related to new name
-   JEditable updated to v2.02
-   Removed dragging of keys
-   Enlarged the playlist box vertically
-   Upgraded to Interact.js v1.3.3
-   Upgraded Howler.js to v2.0.10
-   Updated npm package dependencies

## [0.4.1] - 2017-05-25

### Added

-   Playlist: Dropdown on playlist header to access action items
-   Playlist: Can set initial order via playlist dropdown
-   Playlist: Can reset to initial order via dropdown
-   Playlist: Empty Playlist now has confirmation dialogue

### Fixed

-   Fade times now work properly with 0 seconds
-   Playlist settings no longer stored after dragging/drop sounds within playlist
-   Reduced playlist animation time to improve performance

### Changed

-   Settings menu: Visual redesign
-   Settings menu: Saves when exiting (no need to save explicitly anymore)
-   Updated Sortable.js to v1.5.1

## [0.4.0] - 2017-05-07

### Added

-   Sound Settings: Set/remove 'Played' status
-   General Settings: Mark sounds as played (or not)
-   Pages: Page names and page fade in/out times by right/ctrl-clicking on tab
-   Pages: Scroll on page tabs to get overflow pages
-   Menu: Tutorial option that links to YouTube video

### Fixed

-   Sound settings "cannot read proprty 'name' of undefined" error
-   Deleting sound "cannot read property '0' of undefined" error
-   Deleting sound removes playing and soundNotLoaded classes
-   Waveform shows playhead at proper location after loading waveform

## [0.3.2] - 2017-04-22

### Fixed

-   Playlist: Space key doesn't create error when no sounds in playlist
-   Playlist: Sounds dragged out of top spot no longer stay yellow
-   Playlist: Search doc error fixed

### Changed

-   Opening sound options no longer loads the waveform
-   Playlist: new sounds are added above already played sounds
-   Playlist: When loading multiple sounds, waveform of first sound shows
-   Keyboard: When loading multiple sounds, waveform of first sound shows

## [0.3.1] - 2017-04-08

### Added

-   OS X build

### Changed

-   Updated interact.js to v1.2.8 (solved darwin issues)
-   Backspace also deletes keys (solved darwin issues)

### Fixed

-   Removed error when clicking on key with no sound
-   Loading bar is removed if 0 sounds
-   Copy/paste on darwin
-   Temp fix for random waveformedInfo seek error after ctrl/cmd+X
-   Pasting a song removes 'played' status

## [0.3.0] - 2017-04-07

### Added

-   Pause/stop option in General Settings
-   Global Fade in and Fade out times
-   Pressing 'ENTER' in playlist search un-focuses it
-   Double click on waveform to play sound from that spot
-   COPY/CUT/PASTE shortcuts to move keys around keyboards/pages
-   Import multiple sounds at a time
-   Mac build scripts
-   Icons/background for Mac builds
-   Page names have associated hotkey in name
-   Sound resets after changing waveform region

### Changed

-   License in "About" to Apache 2.0
-   Order of menu items
-   Sounds now "pause" by default (not stop)
-   Minor color/visual fixes
-   Removed start/end from sound settings
-   Can select unloaded keys
-   Cannot launch sounds when holding CONTROL
-   howls are not saved in json - fixed click-key-before-anything-else tracking weirdness (was because playState wasn't always null on launch)
-   Switched to Howler.js for sound engine
-   Removed looping, for now
-   Added text-shadow to emphasize key text
-   Updated Materialize to v0.98.1
-   Updated jQuery to v3.2.1
-   Updated Wavesurfer to v1.3.7
-   Updated Electron to 1.6.4

### Fixed

-   Settings now saving properly
-   Sounds are marked as played when they end
-   Playlist selects first sound after last one stops
-   Empty playlist creates empty object and saves it to the json
-   Can have periods and square brackets in sound names
-   Sounds that don't load are correctly styled
-   Deleting a song no longer breaks waveform sound name
-   Deleting a song removed played class on key
-   Empty playlist search error fixed

data/keyCodes.js
var keyboardMap = [
	"", // [0]
	"", // [1]
	"", // [2]
	"CANCEL", // [3]
	"", // [4]
	"", // [5]
	"HELP", // [6]
	"", // [7]
	"BACK_SPACE", // [8]
	"TAB", // [9]
	"", // [10]
	"", // [11]
	"CLEAR", // [12]
	"ENTER", // [13]
	"ENTER_SPECIAL", // [14]
	"", // [15]
	"SHIFT", // [16]
	"CONTROL", // [17]
	"ALT", // [18]
	"PAUSE", // [19]
	"CAPS_LOCK", // [20]
	"KANA", // [21]
	"EISU", // [22]
	"JUNJA", // [23]
	"FINAL", // [24]
	"HANJA", // [25]
	"", // [26]
	"ESCAPE", // [27]
	"CONVERT", // [28]
	"NONCONVERT", // [29]
	"ACCEPT", // [30]
	"MODECHANGE", // [31]
	"SPACE", // [32]
	"PAGE_UP", // [33]
	"PAGE_DOWN", // [34]
	"END", // [35]
	"HOME", // [36]
	"LEFT", // [37]
	"UP", // [38]
	"RIGHT", // [39]
	"DOWN", // [40]
	"SELECT", // [41]
	"PRINT", // [42]
	"EXECUTE", // [43]
	"PRINTSCREEN", // [44]
	"INSERT", // [45]
	"DELETE", // [46]
	"", // [47]
	"0", // [48]
	"1", // [49]
	"2", // [50]
	"3", // [51]
	"4", // [52]
	"5", // [53]
	"6", // [54]
	"7", // [55]
	"8", // [56]
	"9", // [57]
	"COLON", // [58]
	"SEMICOLON", // [59]
	"LESS_THAN", // [60]
	"EQUALS", // [61]
	"GREATER_THAN", // [62]
	"QUESTION_MARK", // [63]
	"AT", // [64]
	"A", // [65]
	"B", // [66]
	"C", // [67]
	"D", // [68]
	"E", // [69]
	"F", // [70]
	"G", // [71]
	"H", // [72]
	"I", // [73]
	"J", // [74]
	"K", // [75]
	"L", // [76]
	"M", // [77]
	"N", // [78]
	"O", // [79]
	"P", // [80]
	"Q", // [81]
	"R", // [82]
	"S", // [83]
	"T", // [84]
	"U", // [85]
	"V", // [86]
	"W", // [87]
	"X", // [88]
	"Y", // [89]
	"Z", // [90]
	"OS_KEY", // [91] Windows Key (Windows) or Command Key (Mac)
	"", // [92]
	"CONTEXT_MENU", // [93]
	"", // [94]
	"SLEEP", // [95]
	"NUMPAD0", // [96]
	"NUMPAD1", // [97]
	"NUMPAD2", // [98]
	"NUMPAD3", // [99]
	"NUMPAD4", // [100]
	"NUMPAD5", // [101]
	"NUMPAD6", // [102]
	"NUMPAD7", // [103]
	"NUMPAD8", // [104]
	"NUMPAD9", // [105]
	"MULTIPLY", // [106]
	"ADD", // [107]
	"SEPARATOR", // [108]
	"SUBTRACT", // [109]
	"DECIMAL", // [110]
	"DIVIDE", // [111]
	"F1", // [112]
	"F2", // [113]
	"F3", // [114]
	"F4", // [115]
	"F5", // [116]
	"F6", // [117]
	"F7", // [118]
	"F8", // [119]
	"F9", // [120]
	"F10", // [121]
	"F11", // [122]
	"F12", // [123]
	"F13", // [124]
	"F14", // [125]
	"F15", // [126]
	"F16", // [127]
	"F17", // [128]
	"F18", // [129]
	"F19", // [130]
	"F20", // [131]
	"F21", // [132]
	"F22", // [133]
	"F23", // [134]
	"F24", // [135]
	"", // [136]
	"", // [137]
	"", // [138]
	"", // [139]
	"", // [140]
	"", // [141]
	"", // [142]
	"", // [143]
	"NUM_LOCK", // [144]
	"SCROLL_LOCK", // [145]
	"WIN_OEM_FJ_JISHO", // [146]
	"WIN_OEM_FJ_MASSHOU", // [147]
	"WIN_OEM_FJ_TOUROKU", // [148]
	"WIN_OEM_FJ_LOYA", // [149]
	"WIN_OEM_FJ_ROYA", // [150]
	"", // [151]
	"", // [152]
	"", // [153]
	"", // [154]
	"", // [155]
	"", // [156]
	"", // [157]
	"", // [158]
	"", // [159]
	"CIRCUMFLEX", // [160]
	"EXCLAMATION", // [161]
	"DOUBLE_QUOTE", // [162]
	"HASH", // [163]
	"DOLLAR", // [164]
	"PERCENT", // [165]
	"AMPERSAND", // [166]
	"UNDERSCORE", // [167]
	"OPEN_PAREN", // [168]
	"CLOSE_PAREN", // [169]
	"ASTERISK", // [170]
	"PLUS", // [171]
	"PIPE", // [172]
	"HYPHEN_MINUS", // [173]
	"OPEN_CURLY_BRACKET", // [174]
	"CLOSE_CURLY_BRACKET", // [175]
	"TILDE", // [176]
	"", // [177]
	"", // [178]
	"", // [179]
	"", // [180]
	"VOLUME_MUTE", // [181]
	"VOLUME_DOWN", // [182]
	"VOLUME_UP", // [183]
	"", // [184]
	"", // [185]
	"SEMICOLON", // [186]
	"EQUALS", // [187]
	"COMMA", // [188]
	"MINUS", // [189]
	"PERIOD", // [190]
	"SLASH", // [191]
	"BACK_QUOTE", // [192]
	"", // [193]
	"", // [194]
	"", // [195]
	"", // [196]
	"", // [197]
	"", // [198]
	"", // [199]
	"", // [200]
	"", // [201]
	"", // [202]
	"", // [203]
	"", // [204]
	"", // [205]
	"", // [206]
	"", // [207]
	"", // [208]
	"", // [209]
	"", // [210]
	"", // [211]
	"", // [212]
	"", // [213]
	"", // [214]
	"", // [215]
	"", // [216]
	"", // [217]
	"", // [218]
	"OPEN_BRACKET", // [219]
	"BACK_SLASH", // [220]
	"CLOSE_BRACKET", // [221]
	"QUOTE", // [222]
	"", // [223]
	"META", // [224]
	"ALTGR", // [225]
	"", // [226]
	"WIN_ICO_HELP", // [227]
	"WIN_ICO_00", // [228]
	"", // [229]
	"WIN_ICO_CLEAR", // [230]
	"", // [231]
	"", // [232]
	"WIN_OEM_RESET", // [233]
	"WIN_OEM_JUMP", // [234]
	"WIN_OEM_PA1", // [235]
	"WIN_OEM_PA2", // [236]
	"WIN_OEM_PA3", // [237]
	"WIN_OEM_WSCTRL", // [238]
	"WIN_OEM_CUSEL", // [239]
	"WIN_OEM_ATTN", // [240]
	"WIN_OEM_FINISH", // [241]
	"WIN_OEM_COPY", // [242]
	"WIN_OEM_AUTO", // [243]
	"WIN_OEM_ENLW", // [244]
	"WIN_OEM_BACKTAB", // [245]
	"ATTN", // [246]
	"CRSEL", // [247]
	"EXSEL", // [248]
	"EREOF", // [249]
	"PLAY", // [250]
	"ZOOM", // [251]
	"", // [252]
	"PA1", // [253]
	"WIN_OEM_CLEAR", // [254]
	"" // [255]
];
